### PR TITLE
MBS-11609 more debug info for kubernetes

### PR DESCRIPTION
**Pull request by dsvoronin**
KubeContainer and KubePod abstraction layer from fabric8io client
more info about waiting/pending statuses,
point on device log on disconnects

Examples:

## Not yet assigned to node

```
Getting pods for deployment: android-emulator-28efea9e-6891-4915-ac44-0ae34cd6ee80:
----------------
Pod android-emulator-28efea9e-6891-4915-ac44-0ae34cd6ee80-6ff9vmqww [node=null; pod=Pending(message=Unknown); container=Unknown]
------- Summary: running: 0; pending: 1; other 0  ---------
```

## Basic running state

```
Getting pods for deployment: android-emulator-2cb82e1a-fb8b-4211-a1d4-f7798dee71fa:
----------------
Pod android-emulator-28efea9e-6891-4915-ac44-0ae34cd6ee80-6ff9vmqww [node=avi-teamcity04; pod=Running; container=Running]
------- Summary: running: 1; pending: 0; other 0  ---------
```

## No nodes are available

```
Getting pods for deployment: android-emulator-985173cb-d7ce-4878-99dc-f55e513f33ec:
----------------
Pod android-emulator-985173cb-d7ce-4878-99dc-f55e513f33ec-b8d5qcpw5 [node=null; pod=Pending(message=No nodes are available that match all of the predicates: Insufficient cpu (9), Insufficient memory (1), Insufficient pods (16), MatchNodeSelector (139), NodeUnschedulable (6), PodToleratesNodeTaints (136).); container=Unknown]
------- Summary: running: 0; pending: 1; other 0  ---------

```

## Emulator failed to boot
```
Pod android-emulator-28efea9e-6891-4915-ac44-0ae34cd6ee80-6ff9vmqww can't load device. Disconnect and delete.
Check device logs in artifacts: /home/dsvoronin/Work/avito-android-github/samples/test-runner/build/test-runner/a0dedae2d95a769a968cc8773dcf86a6ca227f3f.local-dsvoronin/ui/devices/10.21.218.147.txt
```

and specified file contains:
```
error: kvm run failed Bad address
RAX=ffff97317fc00000 RBX=0000000000011000 RCX=000000000000001f RDX=0000000000000003
RSI=ffffffffafac5722 RDI=ffffffffafa98365 RBP=ffffa45bc34e7c30 RSP=ffffa45bc34e7c28
R8 =0000000000000004 R9 =000000003149ab74 R10=0000000000000000 R11=0000000000000000
R12=ffffffffafc18020 R13=0000000000000000 R14=ffff97317fc0f150 R15=ffffa45bc34e7f58
RIP=ffffffffaea095be RFL=00010006 [-----P-] CPL=0 II=0 A20=1 SMM=0 HLT=0
```

### CODE REVIEW

File: subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/EmulatorsLogsReporter.kt [hunk 1]
```diff
@@ -1,8 +1,13 @@
...
 import com.avito.runner.service.worker.device.Serial
+import java.io.File
 
 internal interface EmulatorsLogsReporter {
+
     fun reportEmulatorLogs(emulatorName: Serial, log: String)
+
     fun redirectLogcat(emulatorName: Serial, device: Device)
+
+    fun getLogFile(podIp: String): File
```
Comment (sergioboi) [inline subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/EmulatorsLogsReporter.kt:12]: For future changes EmulatorsLogsReporter:
The EmulatorsLogsReporter is a strange thing. It knows too much and I don't understand its' responsibility

Maybe add TODO that will help refactor it

File: subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubeContainer.kt [hunk 2]
```diff
@@ -0,0 +1,84 @@
+package com.avito.android.runner.devices.internal.kubernetes
```
Comment (sergioboi) [inline subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubeContainer.kt:1]: Those classes are good candidates for separate module: `kubernetes-extensions`

File: subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesApi.kt [hunk 3]
```diff
@@ -94,63 +94,48 @@ internal interface KubernetesApi {
             }
         }
 
-        override suspend fun getPods(deploymentName: String): Result<List<Pod>> = Result.tryCatch {
+        private fun failFastOnBadImage(pods: List<KubePod>) {
+            pods.forEach { pod ->
+                val phase = pod.container.phase
+                if (phase is KubeContainer.ContainerPhase.Waiting) {
+                    if (phase.hasProblemsGettingImage()) {
+                        error("Can't create pods for deployment, check image reference: ${phase.message}")
```
Comment (sergioboi) [inline subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesApi.kt:102]: Could we return the answer to the caller and make him do the decision?

File: subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesReservationClient.kt [hunk 4]
```diff
@@ -169,11 +167,19 @@ internal class KubernetesReservationClient(
...
     }
 
-    private suspend fun Pod.initializeDevice(): Result<RemoteDevice> {
-        val podName = metadata.name
+    private suspend fun KubePod.initializeDevice(): Result<RemoteDevice> {
+        val podName = name
         logger.debug("Found new pod: $podName")
         return bootDevice().onFailure { error ->
-            logger.warn("Pod $podName can't load device. Disconnect and delete", error)
+            val message = buildString {
```
Comment (sergioboi) [inline subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesReservationClient.kt:174]: Extract to separate fun

**Review by sergioboi — APPROVED** [review]

File: subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubeContainer.kt (hunk 2 above)
Comment (dsvoronin) [reply subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubeContainer.kt:1]: we have such module, see `:kubernetes`, 
but `KubePod` has some internal logic, it assumes that we always have 1 container per pod

File: subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/EmulatorsLogsReporter.kt (hunk 1 above)
Comment (dsvoronin) [reply subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/EmulatorsLogsReporter.kt:12]: i agree

File: subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesApi.kt (hunk 3 above)
Comment (dsvoronin) [reply subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesApi.kt:102]: yes, that's better

File: subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/EmulatorsLogsReporterImpl.kt [hunk 5]
```diff
@@ -10,10 +10,9 @@ internal class EmulatorsLogsReporterImpl(
 ) : EmulatorsLogsReporter {
 
     override fun reportEmulatorLogs(emulatorName: Serial, log: String) {
-        getFile(
-            dir = File(outputFolder, DEVICES_LOGS),
-            emulatorName = emulatorName.value
-        ).writeText(log)
+        getLogFile(emulatorName.value)
+            .apply { parentFile?.mkdirs() }
+            .writeText(log)
```
Comment (eugene-krivobokov) [inline subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/EmulatorsLogsReporterImpl.kt:15]: Maybe append?
If I'm not mistaken there are no restrictions to get a new restarted pos with IP from previously destroyed.

Comment (dsvoronin) [reply subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/EmulatorsLogsReporterImpl.kt:15]: ye, you probably right, and this is a mistake

File: subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesApi.kt (hunk 3 above)
Comment (dsvoronin) [reply subprojects/test-runner/device-provider/impl/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesApi.kt:102]: It's already returned to a client as a `Result.Failure<>` will make this aspect more readable in next PR

**Pull request merged**